File: app/src/main/java/com/termux/app/openhouse/desktop/ui/OpenHouseDesktopView.java
package com.termux.app.openhouse.desktop.ui;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.termux.R;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public final class OpenHouseDesktopView extends LinearLayout {

    public interface EntryViewReadyCallback {
        void onEntryViewReady(View entryView);
    }

    public interface Callbacks {
        default void onOpen(DesktopUiEntry entry) {}
        default void onEdit(DesktopUiEntry entry) {}
        default void onReorder(List<DesktopUiEntry> orderedEntries, DesktopUiEntry movedEntry, int fromPosition, int toPosition) {}
        default void onMoveToSlot(DesktopUiEntry movedEntry, int fromSlot, int toSlot, boolean targetOccupied, boolean createsNewPage) {}
        default void onPageChanged(int pageIndex, int pageCount) {}
        default void onBlankLongPress() {}
        default void onEditModeChanged(boolean editMode) {}
    }

    private static final int EDGE_NONE = 0;
    private static final int EDGE_LEFT = -1;
    private static final int EDGE_RIGHT = 1;
    private static final long AUTO_PAGE_DELAY_MS = 520L;

    private final DesktopViewPager pager;
    private final DesktopPagerAdapter adapter;
    private final TextView pageIndicator;
    private final Handler handler = new Handler(Looper.getMainLooper());
    private final Runnable autoPageRunnable = new Runnable() {
        @Override
        public void run() {
            runAutoPageStep();
        }
    };
    private final List<DesktopUiEntry> entries = new ArrayList<>();
    private Callbacks callbacks = new Callbacks() {};
    private int columns = 3;
    private int rows = 4;
    private boolean editMode;
    private boolean dragInProgress;
    private int pendingAutoPageDirection = EDGE_NONE;

    public OpenHouseDesktopView(Context context) {
        this(context, null);
    }

    public OpenHouseDesktopView(Context context, AttributeSet attrs) {
        super(context, attrs);
        setOrientation(VERTICAL);
        setClipToPadding(false);

        pager = new DesktopViewPager(context);
        pager.setOffscreenPageLimit(1);
        adapter = new DesktopPagerAdapter();
        pager.setAdapter(adapter);
        pager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                updatePageIndicator();
                callbacks.onPageChanged(position, getPageCount());
            }
        });
        addView(pager, new LayoutParams(LayoutParams.MATCH_PARENT, calculatePagerHeight()));

        pageIndicator = new TextView(context);
        pageIndicator.setGravity(android.view.Gravity.CENTER);
        pageIndicator.setTextColor(ContextCompat.getColor(context, R.color.textSecondary));
        pageIndicator.setTextSize(12);
        LayoutParams indicatorParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        indicatorParams.setMargins(0, dp(8), 0, 0);
        addView(pageIndicator, indicatorParams);
        updatePageIndicator();
    }

    public void setCallbacks(Callbacks callbacks) {
        this.callbacks = callbacks == null ? new Callbacks() {} : callbacks;
    }

    public void setEntries(List<DesktopUiEntry> newEntries) {
        entries.clear();
        entries.addAll(normalizeEntries(newEntries));
        int page = Math.min(pager.getCurrentItem(), Math.max(0, getPageCount() - 1));
        adapter.notifyDataSetChanged();
        pager.setCurrentItem(page, false);
        updatePageIndicator();
        callbacks.onPageChanged(pager.getCurrentItem(), getPageCount());
    }

    public List<DesktopUiEntry> getEntries() {
        return Collections.unmodifiableList(snapshotEntriesBySlot());
    }

    public void setEditMode(boolean editMode) {
        if (this.editMode == editMode) {
            return;
        }
        this.editMode = editMode;
        cancelAutoPaging();
        adapter.notifyDataSetChanged();
        clampCurrentPage();
        updatePageIndicator();
        callbacks.onEditModeChanged(editMode);
    }

    public boolean isEditMode() {
        return editMode;
    }

    public void setGridSize(int columns, int rows) {
        this.columns = Math.max(1, columns);
        this.rows = Math.max(1, rows);
        List<DesktopUiEntry> currentEntries = new ArrayList<>(entries);
        entries.clear();
        entries.addAll(normalizeEntries(currentEntries));
        ViewGroup.LayoutParams params = pager.getLayoutParams();
        params.height = calculatePagerHeight();
        pager.setLayoutParams(params);
        adapter.notifyDataSetChanged();
        clampCurrentPage();
        updatePageIndicator();
    }

    public int getCurrentPage() {
        return pager.getCurrentItem();
    }

    /**
     * Makes the page containing {@code entryId} visible and reports the laid-out tile view.
     * Tutorial callers can then use {@link #findEntryView(String)} as a stable target supplier.
     */
    public boolean revealEntry(String entryId, EntryViewReadyCallback callback) {
        String normalizedId = DesktopUiEntry.safeTrim(entryId);
        DesktopUiEntry entry = findEntry(normalizedId);
        if (entry == null) {
            return false;
        }
        setCurrentPage(entry.slotIndex / getPageSize(), false);
        dispatchEntryViewWhenReady(normalizedId, callback);
        return true;
    }

    public View findEntryView(String entryId) {
        String normalizedId = DesktopUiEntry.safeTrim(entryId);
        if (normalizedId.isEmpty()) {
            return null;
        }
        return findViewWithTag(DesktopAppTileView.entryTag(normalizedId));
    }

    public void setCurrentPage(int page, boolean smoothScroll) {
        int target = Math.max(0, Math.min(page, getPageCount() - 1));
        pager.setCurrentItem(target, smoothScroll);
        updatePageIndicator();
    }

    public int getPageCount() {
        int actualCount = getActualPageCount();
        if (editMode && !entries.isEmpty()) {
            return actualCount + 1;
        }
        return actualCount;
    }

    private int getActualPageCount() {
        int pageSize = getPageSize();
        int maxSlot = -1;
        for (DesktopUiEntry entry : entries) {
            if (entry != null && entry.slotIndex > maxSlot) {
                maxSlot = entry.slotIndex;
            }
        }
        if (maxSlot < 0) {
            return 1;
        }
        return (maxSlot / pageSize) + 1;
    }

    private int getPageSize() {
        return Math.max(1, columns * rows);
    }

    private int calculatePagerHeight() {
        return dp(20) + rows * dp(116) + Math.max(0, rows - 1) * dp(10);
    }

    private void requestEditModeFromGesture() {
        if (!editMode) {
            setEditMode(true);
        }
    }

    private void moveEntryToSlot(String draggedId, int targetSlot) {
        String normalizedId = DesktopUiEntry.safeTrim(draggedId);
        if (normalizedId.isEmpty() || entries.isEmpty()) {
            return;
        }
        int fromIndex = findEntryIndex(normalizedId);
        if (fromIndex < 0) {
            return;
        }
        int safeTargetSlot = Math.max(0, targetSlot);
        DesktopUiEntry moved = entries.get(fromIndex);
        int fromSlot = moved.slotIndex;
        if (fromSlot == safeTargetSlot) {
            return;
        }

        int actualPageCountBeforeMove = getActualPageCount();
        List<DesktopUiEntry> updated = new ArrayList<>();
        boolean targetOccupied = false;
        for (DesktopUiEntry entry : entries) {
            if (entry == null || normalizedId.equals(entry.id)) {
                continue;
            }
            if (entry.slotIndex == safeTargetSlot) {
                targetOccupied = true;
                updated.add(entry.withSlotIndex(fromSlot));
            } else {
                updated.add(entry);
            }
        }

        DesktopUiEntry movedToTarget = moved.withSlotIndex(safeTargetSlot);
        updated.add(movedToTarget);
        entries.clear();
        entries.addAll(normalizeEntries(updated));
        adapter.notifyDataSetChanged();
        setCurrentPage(safeTargetSlot / getPageSize(), false);

        boolean createsNewPage = safeTargetSlot >= actualPageCountBeforeMove * getPageSize();
        callbacks.onMoveToSlot(movedToTarget, fromSlot, safeTargetSlot, targetOccupied, createsNewPage);
        callbacks.onReorder(Collections.unmodifiableList(snapshotEntriesBySlot()), movedToTarget, fromSlot, safeTargetSlot);
    }

    private int findEntryIndex(String id) {
        String normalized = DesktopUiEntry.safeTrim(id);
        for (int i = 0; i < entries.size(); i++) {
            if (normalized.equals(entries.get(i).id)) {
                return i;
            }
        }
        return -1;
    }

    private DesktopUiEntry findEntry(String id) {
        int index = findEntryIndex(id);
        return index < 0 ? null : entries.get(index);
    }

    private void dispatchEntryViewWhenReady(String entryId, EntryViewReadyCallback callback) {
        if (callback == null) {
            return;
        }
        pager.post(() -> {
            View entryView = findEntryView(entryId);
            if (isReadyTarget(entryView)) {
                callback.onEntryViewReady(entryView);
                return;
            }
            getViewTreeObserver().addOnGlobalLayoutListener(new android.view.ViewTreeObserver.OnGlobalLayoutListener() {
                @Override
                public void onGlobalLayout() {
                    View laidOutEntry = findEntryView(entryId);
                    if (!isReadyTarget(laidOutEntry)) {
                        return;
                    }
                    if (getViewTreeObserver().isAlive()) {
                        getViewTreeObserver().removeOnGlobalLayoutListener(this);
                    }
                    callback.onEntryViewReady(laidOutEntry);
                }
            });
        });
    }

    private boolean isReadyTarget(View view) {
        return view != null && view.getWidth() > 0 && view.getHeight() > 0 && view.isShown();
    }

    private List<DesktopUiEntry> normalizeEntries(List<DesktopUiEntry> source) {
        if (source == null || source.isEmpty()) {
            return Collections.emptyList();
        }

        List<EntrySeed> positionedSeeds = new ArrayList<>();
        List<EntrySeed> implicitSeeds = new ArrayList<>();
        int inputIndex = 0;
        for (DesktopUiEntry entry : source) {
            if (entry == null || DesktopUiEntry.safeTrim(entry.id).isEmpty()) {
                inputIndex++;
                continue;
            }
            EntrySeed seed = new EntrySeed(entry, inputIndex);
            if (entry.hasExplicitSlotIndex() || entry.hasExplicitOrder()) {
                positionedSeeds.add(seed);
            } else {
                implicitSeeds.add(seed);
            }
            inputIndex++;
        }

        Collections.sort(positionedSeeds, (left, right) -> {
            int slotCompare = Integer.compare(left.entry.requestedSlotIndex(left.inputIndex), right.entry.requestedSlotIndex(right.inputIndex));
            return slotCompare != 0 ? slotCompare : Integer.compare(left.inputIndex, right.inputIndex);
        });
        Collections.sort(implicitSeeds, (left, right) -> {
            int leftOrder = left.entry.hasExplicitOrder() ? left.entry.order : left.inputIndex;
            int rightOrder = right.entry.hasExplicitOrder() ? right.entry.order : right.inputIndex;
            int orderCompare = Integer.compare(leftOrder, rightOrder);
            return orderCompare != 0 ? orderCompare : Integer.compare(left.inputIndex, right.inputIndex);
        });

        List<DesktopUiEntry> placed = new ArrayList<>();
        Set<Integer> usedSlots = new HashSet<>();
        List<EntrySeed> overflow = new ArrayList<>();
        for (EntrySeed seed : positionedSeeds) {
            int slot = seed.entry.requestedSlotIndex(seed.inputIndex);
            if (usedSlots.add(slot)) {
                placed.add(seed.entry.withSlotIndex(slot));
            } else {
                overflow.add(seed);
            }
        }
        overflow.addAll(implicitSeeds);
        int nextFreeSlot = 0;
        for (EntrySeed seed : overflow) {
            while (usedSlots.contains(nextFreeSlot)) {
                nextFreeSlot++;
            }
            usedSlots.add(nextFreeSlot);
            placed.add(seed.entry.withSlotIndex(nextFreeSlot));
            nextFreeSlot++;
        }
        Collections.sort(placed, (left, right) -> Integer.compare(left.slotIndex, right.slotIndex));
        return placed;
    }

    private List<DesktopUiEntry> snapshotEntriesBySlot() {
        List<DesktopUiEntry> snapshot = new ArrayList<>(entries);
        Collections.sort(snapshot, (left, right) -> Integer.compare(left.slotIndex, right.slotIndex));
        return snapshot;
    }

    private List<DesktopUiEntry> buildPageSlots(int pageIndex) {
        int pageSize = getPageSize();
        int baseSlot = Math.max(0, pageIndex) * pageSize;
        List<DesktopUiEntry> slots = new ArrayList<>(Collections.nCopies(pageSize, null));
        for (DesktopUiEntry entry : entries) {
            if (entry == null) {
                continue;
            }
            int offset = entry.slotIndex - baseSlot;
            if (offset >= 0 && offset < pageSize) {
                slots.set(offset, entry);
            }
        }
        return slots;
    }

    private void handleDragStarted() {
        dragInProgress = true;
        pager.setDragInProgress(true);
    }

    private void handleDragEnded() {
        dragInProgress = false;
        pager.setDragInProgress(false);
        cancelAutoPaging();
    }

    private void handleDragLocation(float rawX) {
        if (!editMode || !dragInProgress || pager.getWidth() <= 0) {
            cancelAutoPaging();
            return;
        }
        int[] pagerLocation = new int[2];
        pager.getLocationOnScreen(pagerLocation);
        int edgeSize = dp(54);
        int pagerLeft = pagerLocation[0];
        int pagerRight = pagerLeft + pager.getWidth();
        if (rawX <= pagerLeft + edgeSize) {
            scheduleAutoPaging(EDGE_LEFT);
        } else if (rawX >= pagerRight - edgeSize) {
            scheduleAutoPaging(EDGE_RIGHT);
        } else {
            cancelAutoPaging();
        }
    }

    private void scheduleAutoPaging(int direction) {
        if (direction == EDGE_NONE) {
            cancelAutoPaging();
            return;
        }
        if (pendingAutoPageDirection == direction) {
            return;
        }
        cancelAutoPaging();
        pendingAutoPageDirection = direction;
        handler.postDelayed(autoPageRunnable, AUTO_PAGE_DELAY_MS);
    }

    private void runAutoPageStep() {
        int direction = pendingAutoPageDirection;
        if (direction == EDGE_NONE || !dragInProgress) {
            cancelAutoPaging();
            return;
        }
        int current = pager.getCurrentItem();
        int target = current + direction;
        if (target >= 0 && target < getPageCount()) {
            pager.setCurrentItem(target, true);
            handler.postDelayed(autoPageRunnable, AUTO_PAGE_DELAY_MS);
        } else {
            cancelAutoPaging();
        }
    }

    private void cancelAutoPaging() {
        pendingAutoPageDirection = EDGE_NONE;
        handler.removeCallbacks(autoPageRunnable);
    }

    private void clampCurrentPage() {
        int target = Math.max(0, Math.min(pager.getCurrentItem(), getPageCount() - 1));
        if (target != pager.getCurrentItem()) {
            pager.setCurrentItem(target, false);
        }
    }

    private void updatePageIndicator() {
        int pageCount = getPageCount();
        pageIndicator.setText((pager.getCurrentItem() + 1) + " / " + pageCount);
        pageIndicator.setVisibility(pageCount > 1 ? View.VISIBLE : View.INVISIBLE);
    }

    private int dp(int value) {
        float density = getResources().getDisplayMetrics().density;
        return Math.round(value * density);
    }

    private static final class EntrySeed {
        final DesktopUiEntry entry;
        final int inputIndex;

        EntrySeed(DesktopUiEntry entry, int inputIndex) {
            this.entry = entry;
            this.inputIndex = inputIndex;
        }
    }

    private final class DesktopPagerAdapter extends PagerAdapter {

        @Override
        public int getCount() {
            return getPageCount();
        }

        @Override
        public int getItemPosition(@NonNull Object object) {
            return POSITION_NONE;
        }

        @Override
        public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
            return view == object;
        }

        @NonNull
        @Override
        public Object instantiateItem(@NonNull ViewGroup container, int position) {
            int pageSize = getPageSize();
            int baseSlot = position * pageSize;
            DesktopPageView page = new DesktopPageView(container.getContext());
            page.bind(buildPageSlots(position), position, baseSlot, columns, rows, editMode, new DesktopPageView.Callback() {
                @Override
                public void onOpen(DesktopUiEntry entry) {
                    callbacks.onOpen(entry);
                }

                @Override
                public void onEdit(DesktopUiEntry entry) {
                    callbacks.onEdit(entry);
                }

                @Override
                public void onRequestEditMode(DesktopUiEntry entry) {
                    requestEditModeFromGesture();
                }

                @Override
                public void onDragStarted(DesktopDragPayload payload) {
                    handleDragStarted();
                }

                @Override
                public void onDragEnded() {
                    handleDragEnded();
                }

                @Override
                public void onDragLocation(float rawX) {
                    handleDragLocation(rawX);
                }

                @Override
                public void onMove(String draggedId, int targetSlot) {
                    moveEntryToSlot(draggedId, targetSlot);
                }

                @Override
                public void onBlankLongPress() {
                    requestEditModeFromGesture();
                    callbacks.onBlankLongPress();
                }
            });
            container.addView(page);
            return page;
        }

        @Override
        public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
            container.removeView((View) object);
        }
    }

    private static final class DesktopViewPager extends ViewPager {

        private boolean dragInProgress;

        DesktopViewPager(Context context) {
            super(context);
        }

        void setDragInProgress(boolean dragInProgress) {
            this.dragInProgress = dragInProgress;
            requestDisallowInterceptTouchEvent(dragInProgress);
        }

        @Override
        public boolean onInterceptTouchEvent(android.view.MotionEvent ev) {
            return !dragInProgress && super.onInterceptTouchEvent(ev);
        }

        @Override
        public boolean onTouchEvent(android.view.MotionEvent ev) {
            return !dragInProgress && super.onTouchEvent(ev);
        }
    }
}

File: openhouse-feature/src/main/java/com/wuxianpi/openhouse/feature/desktop/ui/OpenHouseDesktopView.kt
package com.wuxianpi.openhouse.feature.desktop.ui

import android.content.ClipData
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.AttributeSet
import android.view.DragEvent
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager
import com.wuxianpi.openhouse.feature.R
import com.wuxianpi.openhouse.feature.desktop.DesktopLayoutEntry
import java.util.Collections

data class DesktopUiEntry(
    val id: String,
    val title: String,
    val subtitle: String,
    val iconLabel: String,
    val iconKey: String,
    val slotIndex: Int,
    val enabled: Boolean,
) {
    fun displayTitle(): String = title.ifBlank { id }

    fun displayIconLabel(): String {
        val text = iconLabel.ifBlank { iconKey.ifBlank { displayTitle() } }
        return if (text.codePointCount(0, text.length) <= 2) text
        else text.substring(0, text.offsetByCodePoints(0, 1))
    }

    companion object {
        fun from(entry: DesktopLayoutEntry) = DesktopUiEntry(
            id = entry.id,
            title = entry.title,
            subtitle = entry.subtitle,
            iconLabel = entry.iconLabel,
            iconKey = entry.iconKey,
            slotIndex = entry.slotIndex,
            enabled = entry.component.enabled,
        )
    }
}

class OpenHouseDesktopView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
) : LinearLayout(context, attrs) {
    interface Callbacks {
        fun onOpen(entry: DesktopUiEntry) = Unit
        fun onEdit(entry: DesktopUiEntry) = Unit
        fun onMove(entry: DesktopUiEntry, fromSlot: Int, toSlot: Int) = Unit
        fun onPageChanged(pageIndex: Int, pageCount: Int) = Unit
        fun onBlankLongPress() = Unit
        fun onEditModeChanged(editMode: Boolean) = Unit
    }

    private val pager = DesktopViewPager(context)
    private val adapter = DesktopPagerAdapter()
    private val indicator = TextView(context)
    private val entries = mutableListOf<DesktopUiEntry>()
    private val handler = Handler(Looper.getMainLooper())
    private var callbacks: Callbacks = object : Callbacks {}
    private var columns = 3
    private var rows = 4
    private var editMode = false
    private var dragInProgress = false
    private var pendingPageDirection = 0
    private val autoPageTask = object : Runnable {
        override fun run() {
            if (!dragInProgress || pendingPageDirection == 0) return
            val target = pager.currentItem + pendingPageDirection
            if (target in 0 until pageCount()) {
                pager.setCurrentItem(target, true)
                handler.postDelayed(this, AUTO_PAGE_DELAY_MS)
            } else {
                cancelAutoPage()
            }
        }
    }

    init {
        orientation = VERTICAL
        clipToPadding = false
        pager.adapter = adapter
        pager.offscreenPageLimit = 1
        pager.addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
            override fun onPageSelected(position: Int) {
                updateIndicator()
                callbacks.onPageChanged(position, pageCount())
            }
        })
        addView(pager, LayoutParams(LayoutParams.MATCH_PARENT, pagerHeight()))

        indicator.gravity = Gravity.CENTER
        indicator.setTextColor(ContextCompat.getColor(context, R.color.oh_text_secondary))
        indicator.textSize = 12f
        addView(indicator, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT).apply {
            topMargin = dp(8)
        })
        updateIndicator()
    }

    fun setCallbacks(value: Callbacks?) {
        callbacks = value ?: object : Callbacks {}
    }

    fun setEntries(value: List<DesktopUiEntry>?) {
        entries.clear()
        entries += normalize(value.orEmpty())
        adapter.notifyDataSetChanged()
        setCurrentPage(pager.currentItem, false)
        updateIndicator()
    }

    fun getEntries(): List<DesktopUiEntry> = entries.sortedBy { it.slotIndex }

    fun setGridSize(columns: Int, rows: Int) {
        this.columns = columns.coerceAtLeast(1)
        this.rows = rows.coerceAtLeast(1)
        pager.layoutParams = pager.layoutParams.apply { height = pagerHeight() }
        adapter.notifyDataSetChanged()
        setCurrentPage(pager.currentItem, false)
    }

    fun setEditMode(enabled: Boolean) {
        if (editMode == enabled) return
        editMode = enabled
        cancelAutoPage()
        adapter.notifyDataSetChanged()
        setCurrentPage(pager.currentItem, false)
        callbacks.onEditModeChanged(enabled)
    }

    fun isEditMode(): Boolean = editMode

    fun currentPage(): Int = pager.currentItem

    fun setCurrentPage(page: Int, smooth: Boolean) {
        pager.setCurrentItem(page.coerceIn(0, pageCount() - 1), smooth)
        updateIndicator()
    }

    fun releaseResources() {
        cancelAutoPage()
        dragInProgress = false
        callbacks = object : Callbacks {}
        entries.clear()
        pager.adapter = null
        removeAllViews()
    }

    private fun normalize(source: List<DesktopUiEntry>): List<DesktopUiEntry> {
        val used = mutableSetOf<Int>()
        val result = mutableListOf<DesktopUiEntry>()
        source.filter { it.id.isNotBlank() }.sortedBy { it.slotIndex }.forEach { entry ->
            var slot = entry.slotIndex.coerceAtLeast(0)
            while (slot in used) slot++
            used += slot
            result += entry.copy(slotIndex = slot)
        }
        return result
    }

    private fun pageCount(): Int {
        val actual = ((entries.maxOfOrNull { it.slotIndex } ?: -1) / pageSize()) + 1
        return (if (editMode && entries.isNotEmpty()) actual + 1 else actual).coerceAtLeast(1)
    }

    private fun pageEntries(page: Int): List<DesktopUiEntry?> {
        val base = page * pageSize()
        return MutableList<DesktopUiEntry?>(pageSize()) { null }.also { slots ->
            entries.forEach { entry ->
                val offset = entry.slotIndex - base
                if (offset in slots.indices) slots[offset] = entry
            }
        }
    }

    private fun move(entryId: String, targetSlot: Int) {
        val index = entries.indexOfFirst { it.id == entryId }
        if (index < 0) return
        val moved = entries[index]
        val from = moved.slotIndex
        val target = targetSlot.coerceAtLeast(0)
        if (from == target) return
        val occupantIndex = entries.indexOfFirst { it.slotIndex == target && it.id != entryId }
        if (occupantIndex >= 0) entries[occupantIndex] = entries[occupantIndex].copy(slotIndex = from)
        entries[index] = moved.copy(slotIndex = target)
        entries.sortBy { it.slotIndex }
        adapter.notifyDataSetChanged()
        setCurrentPage(target / pageSize(), false)
        callbacks.onMove(entries.first { it.id == entryId }, from, target)
    }

    private fun onDragLocation(rawX: Float) {
        if (!editMode || !dragInProgress || pager.width <= 0) return cancelAutoPage()
        val location = IntArray(2)
        pager.getLocationOnScreen(location)
        val edge = dp(54)
        val direction = when {
            rawX <= location[0] + edge -> -1
            rawX >= location[0] + pager.width - edge -> 1
            else -> 0
        }
        if (direction == 0) return cancelAutoPage()
        if (pendingPageDirection == direction) return
        cancelAutoPage()
        pendingPageDirection = direction
        handler.postDelayed(autoPageTask, AUTO_PAGE_DELAY_MS)
    }

    private fun cancelAutoPage() {
        pendingPageDirection = 0
        handler.removeCallbacks(autoPageTask)
    }

    private fun updateIndicator() {
        val count = pageCount()
        indicator.text = "${pager.currentItem + 1} / $count"
        indicator.visibility = if (count > 1) View.VISIBLE else View.INVISIBLE
    }

    private fun pageSize() = columns * rows
    private fun pagerHeight() = dp(20) + rows * dp(116) + (rows - 1) * dp(10)
    private fun dp(value: Int) = (value * resources.displayMetrics.density).toInt()

    private inner class DesktopPagerAdapter : PagerAdapter() {
        override fun getCount(): Int = pageCount()
        override fun isViewFromObject(view: View, item: Any): Boolean = view === item
        override fun getItemPosition(item: Any): Int = POSITION_NONE

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val page = DesktopPageView(container.context)
            page.bind(pageEntries(position), position * pageSize(), columns, rows, editMode, object : PageCallbacks {
                override fun open(entry: DesktopUiEntry) = callbacks.onOpen(entry)
                override fun edit(entry: DesktopUiEntry) = callbacks.onEdit(entry)
                override fun requestEdit(entry: DesktopUiEntry?) {
                    if (!editMode) setEditMode(true)
                    if (entry == null) callbacks.onBlankLongPress()
                }
                override fun dragStarted() {
                    dragInProgress = true
                    pager.dragInProgress = true
                }
                override fun dragEnded() {
                    dragInProgress = false
                    pager.dragInProgress = false
                    cancelAutoPage()
                }
                override fun dragLocation(rawX: Float) = onDragLocation(rawX)
                override fun move(entryId: String, targetSlot: Int) {
                    this@OpenHouseDesktopView.move(entryId, targetSlot)
                }
            })
            container.addView(page)
            return page
        }

        override fun destroyItem(container: ViewGroup, position: Int, item: Any) {
            container.removeView(item as View)
        }
    }

    private class DesktopViewPager(context: Context) : ViewPager(context) {
        var dragInProgress = false
            set(value) {
                field = value
                requestDisallowInterceptTouchEvent(value)
            }

        override fun onInterceptTouchEvent(event: android.view.MotionEvent): Boolean =
            !dragInProgress && super.onInterceptTouchEvent(event)

        override fun onTouchEvent(event: android.view.MotionEvent): Boolean =
            !dragInProgress && super.onTouchEvent(event)
    }

    companion object {
        private const val AUTO_PAGE_DELAY_MS = 520L
    }
}

private interface PageCallbacks {
    fun open(entry: DesktopUiEntry)
    fun edit(entry: DesktopUiEntry)
    fun requestEdit(entry: DesktopUiEntry?)
    fun dragStarted()
    fun dragEnded()
    fun dragLocation(rawX: Float)
    fun move(entryId: String, targetSlot: Int)
}

private data class DragPayload(val entryId: String, val fromSlot: Int)

private class DesktopPageView(context: Context) : FrameLayout(context) {
    private val grid = LinearLayout(context).apply {
        orientation = LinearLayout.VERTICAL
        gravity = Gravity.CENTER
    }
    private var callback: PageCallbacks? = null
    private var editMode = false
    private var baseSlot = 0

    init {
        setPadding(dp(10), dp(8), dp(10), dp(8))
        addView(grid, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        setOnLongClickListener {
            callback?.requestEdit(null)
            callback != null
        }
    }

    fun bind(
        entries: List<DesktopUiEntry?>,
        baseSlot: Int,
        columns: Int,
        rows: Int,
        editMode: Boolean,
        callback: PageCallbacks,
    ) {
        this.callback = callback
        this.editMode = editMode
        this.baseSlot = baseSlot
        grid.removeAllViews()
        var index = 0
        repeat(rows) { rowIndex ->
            val row = LinearLayout(context).apply {
                orientation = LinearLayout.HORIZONTAL
                gravity = Gravity.CENTER
            }
            grid.addView(row, LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT, 0, 1f).apply {
                if (rowIndex > 0) topMargin = dp(10)
            })
            repeat(columns) {
                val entry = entries.getOrNull(index)
                val slot = baseSlot + index
                val child = if (entry == null) blankSlot(slot) else tile(entry, slot)
                row.addView(child, LinearLayout.LayoutParams(0, LayoutParams.MATCH_PARENT, 1f).apply {
                    leftMargin = dp(7)
                    rightMargin = dp(7)
                })
                index++
            }
        }
    }

    private fun tile(entry: DesktopUiEntry, slot: Int): View = DesktopTileView(context).apply {
        bind(entry, slot, editMode, callback!!)
        setOnDragListener { view, event -> handleDrag(view, event, slot) }
    }

    private fun blankSlot(slot: Int): View = FrameLayout(context).apply {
        minimumHeight = dp(104)
        alpha = if (editMode) 1f else 0f
        background = if (editMode) blankBackground(false) else null
        setOnLongClickListener {
            callback?.requestEdit(null)
            callback != null
        }
        setOnDragListener { view, event -> handleDrag(view, event, slot) }
    }

    private fun handleDrag(target: View, event: DragEvent, targetSlot: Int): Boolean {
        val payload = event.localState as? DragPayload ?: return false
        if (!editMode) return event.action == DragEvent.ACTION_DRAG_STARTED
        return when (event.action) {
            DragEvent.ACTION_DRAG_STARTED -> true
            DragEvent.ACTION_DRAG_LOCATION -> {
                val location = IntArray(2)
                target.getLocationOnScreen(location)
                callback?.dragLocation(location[0] + event.x)
                true
            }
            DragEvent.ACTION_DRAG_ENTERED -> { highlight(target, true); true }
            DragEvent.ACTION_DRAG_EXITED -> { highlight(target, false); true }
            DragEvent.ACTION_DROP -> {
                highlight(target, false)
                callback?.move(payload.entryId, targetSlot)
                true
            }
            DragEvent.ACTION_DRAG_ENDED -> {
                highlight(target, false)
                callback?.dragEnded()
                true
            }
            else -> true
        }
    }

    private fun highlight(view: View, highlighted: Boolean) {
        val scale = if (highlighted) 1.045f else 1f
        view.scaleX = scale
        view.scaleY = scale
        if (view is FrameLayout && view !is DesktopTileView) {
            view.background = if (editMode) blankBackground(highlighted) else null
        }
    }

    private fun blankBackground(highlighted: Boolean) = GradientDrawable().apply {
        shape = GradientDrawable.RECTANGLE
        cornerRadius = dp(6).toFloat()
        setColor(if (highlighted) Color.argb(38, 20, 125, 100) else Color.argb(12, 80, 80, 80))
        setStroke(dp(1), if (highlighted) Color.argb(150, 20, 125, 100) else Color.argb(60, 80, 80, 80))
    }

    private fun dp(value: Int) = (value * resources.displayMetrics.density).toInt()
}

private class DesktopTileView(context: Context) : FrameLayout(context) {
    private val column = LinearLayout(context).apply {
        orientation = LinearLayout.VERTICAL
        gravity = Gravity.CENTER
        setPadding(dp(6), dp(8), dp(6), dp(7))
    }
    private val icon = TextView(context).apply {
        gravity = Gravity.CENTER
        setTextColor(ContextCompat.getColor(context, R.color.oh_text))
        setTypeface(typeface, android.graphics.Typeface.BOLD)
        setBackgroundResource(R.drawable.oh_icon_background)
    }
    private val title = TextView(context).apply {
        gravity = Gravity.CENTER
        setTextColor(ContextCompat.getColor(context, R.color.oh_text))
        textSize = 12f
        maxLines = 2
    }

    init {
        setBackgroundResource(R.drawable.oh_button_background)
        isClickable = true
        isFocusable = true
        minimumHeight = dp(104)
        addView(column, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
        column.addView(icon, LinearLayout.LayoutParams(dp(48), dp(48)))
        column.addView(title, LinearLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT).apply {
            topMargin = dp(7)
        })
    }

    fun bind(entry: DesktopUiEntry, slot: Int, editMode: Boolean, callback: PageCallbacks) {
        icon.text = entry.displayIconLabel()
        val iconText = icon.text.toString()
        icon.textSize = if (iconText.codePointCount(0, iconText.length) > 1) 19f else 24f
        title.text = entry.displayTitle()
        isEnabled = entry.enabled
        alpha = if (entry.enabled) 1f else .55f
        setOnClickListener { if (editMode) callback.edit(entry) else callback.open(entry) }
        setOnLongClickListener {
            if (!editMode) {
                callback.requestEdit(entry)
                true
            } else {
                val payload = DragPayload(entry.id, slot)
                val data = ClipData.newPlainText("openhouse-desktop-entry", entry.id)
                val started = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    startDragAndDrop(data, DragShadowBuilder(this), payload, 0)
                } else {
                    @Suppress("DEPRECATION")
                    startDrag(data, DragShadowBuilder(this), payload, 0)
                }
                if (started) callback.dragStarted() else callback.dragEnded()
                started
            }
        }
    }

    private fun dp(value: Int) = (value * resources.displayMetrics.density).toInt()
}
